### PR TITLE
[mccall_model] JAX conversion of the Job Search Model I

### PR DESCRIPTION
**Pull request by longye-tian**
I've updated the Job Search Model I, align with the change we've had before on Job Search Model II and III. 

## Overview
Converted `lectures/mccall_model.md` from NumPy/Numba to JAX implementation, following the patterns established in `mccall_model_with_separation.md` and `mccall_fitted_vfi.md`.

### 1. Library Updates
- **Replaced:** `numpy` → `jax.numpy` (jnp)
- **Replaced:** `numba.jit` → `jax.jit`
- **Removed:** `numba.experimental.jitclass`
- **Added:** `jax`, `jax.numpy`, `typing.NamedTuple`

### 2. Data Structure Changes
- **Before:** Used Numba's `@jitclass` for the McCall model class
- **After:** Used `NamedTuple` for immutable model structure, compatible with JAX's functional paradigm

### 3. Array Operations
- **Converted:** All NumPy arrays to JAX arrays (jnp)
- **Updated:** Array mutations to use JAX's functional update syntax (`.at[].set()`)

### 4. Loop Transformations
- **Before:** Standard Python while loops with Numba JIT
- **After:** `jax.lax.while_loop` with condition and body functions

### 5. Random Number Generation
- **Before:** NumPy's global random state (`np.random.seed()`)
- **After:** JAX's explicit PRNG key system (`jax.random.PRNGKey`)

### 6. Vectorization
- **Before:** Manual loops over parameter grids
- **After:** `jax.vmap` for automatic vectorization

### CODE REVIEW

**Comment by HumphreyYang** [comment]
Many thanks @longye-tian!

I thought @shlff self-assigned this lecture in the issue here: https://github.com/QuantEcon/meta/issues/173.

Just CCing @shlff here so we are not working on the same lecture separately : )

**Comment by longye-tian** [comment]
> Many thanks @longye-tian!
> 
> I thought @shlff self-assigned this lecture in the issue here: [QuantEcon/meta#173](https://github.com/QuantEcon/meta/issues/173).
> 
> Just CCing @shlff here so we are not working on the same lecture separately : )

Thank you for your notice @HumphreyYang .

So Sorry @shlff . I forgot to check this list...

File: lectures/mccall_model.md [hunk 1]
```diff
@@ -62,8 +62,9 @@ Let's start with some imports:
 ```{code-cell} ipython
```
Comment (HumphreyYang) [inline lectures/mccall_model.md:62]: Before imports we should add `jax` install:

````
```{code-cell} ipython3
:tags: [hide-output]

!pip install quantecon jax
```
````

File: lectures/mccall_model.md [hunk 2]
```diff
@@ -343,14 +344,14 @@ Our default for $q$, the distribution of the state process, will be
 
 ```{code-cell} python3
 n, a, b = 50, 200, 100                        # default parameters
-q_default = BetaBinomial(n, a, b).pdf()       # default choice of q
+q_default = jnp.array(BetaBinomial(n, a, b).pdf())  # default choice of q as JAX array
```
Comment (HumphreyYang) [inline lectures/mccall_model.md]: Just trying to keep it within 80 characters:

```suggestion
q_default = jnp.array(BetaBinomial(n, a, b).pdf())
```

File: lectures/mccall_model.md [hunk 3]
```diff
@@ -364,32 +365,10 @@ ax.set_ylabel('probabilities')
...
-
-This notation specifies that the arrays should be C-contiguous.
-
-This is important for performance, especially when using the `@` operator for matrix multiplication (e.g., `v @ q`). 
-
-Without this specification, Numba might need to handle non-contiguous arrays, which can significantly slow down these operations.
-
-Try to replace `[::1]` with `[:]` and see what happens.
-```
+* JAX provides automatic differentiation and JIT compilation capabilities.
```
Comment (HumphreyYang) [inline lectures/mccall_model.md]: It's a nice motivating paragraph, but just curious if we need this discussion in this lecture since we have JAX lecture showing these already 🤔

We are also not using the autodiff here : )

File: lectures/mccall_model.md [hunk 4]
```diff
@@ -364,32 +365,10 @@ ax.set_ylabel('probabilities')
...
+* We'll use NamedTuple for our model class to maintain immutability, which works well with JAX's functional programming paradigm.
 
 Here's a class that stores the data and computes the values of state-action pairs,
```
Comment (HumphreyYang) [inline lectures/mccall_model.md]: "Here's a class that stores the data and computes the values of state-action pairs,"

This discussion is outdated because the class is only used to store data and the computation of the value is done in `state_action_values`

File: lectures/mccall_model.md [hunk 5]
```diff
@@ -398,26 +377,25 @@ given the current state and an arbitrary feasible action.
...
-        accept = w[i] / (1 - β)
-        reject = c + β * (v @ q)
-
-        return np.array([accept, reject])
+class McCallModel(NamedTuple):
+    c: float = 25           # unemployment compensation
+    β: float = 0.99         # discount factor  
+    w: jnp.ndarray = w_default  # array of wage values, w[i] = wage at state i
+    q: jnp.ndarray = q_default  # array of probabilities
+
```
Comment (HumphreyYang) [inline lectures/mccall_model.md:380]:
```suggestion
class McCallModel(NamedTuple):
    c: float = 25               # unemployment compensation
    β: float = 0.99             # discount factor  
    w: jnp.ndarray = w_default  # array of wage values, w[i] = wage at state i
    q: jnp.ndarray = q_default  # array of probabilities

```

File: lectures/mccall_model.md [hunk 6]
```diff
@@ -678,37 +654,42 @@ Plot mean unemployment duration as a function of $c$ in `c_vals`.
...
+    def cond_fun(state):
+        t, _, done = state
+        return jnp.logical_not(done)
+    
+    initial_state = (1, key, False)
+    t_final, _, _ = jax.lax.while_loop(cond_fun, body_fun, initial_state)
+    return t_final
+
+@jax.jit
+def compute_mean_stopping_time(w_bar, num_reps=100000, seed=1234):
```
Comment (HumphreyYang) [inline lectures/mccall_model.md:678]: Let's use `seed=0` by default as suggested by John.

```suggestion
def compute_mean_stopping_time(w_bar, num_reps=100000, seed=0):
```

File: lectures/mccall_model.md [hunk 7]
```diff
@@ -789,48 +770,41 @@ Once your code is working, investigate how the reservation wage changes with $c$
...
-
-
-@jit
+class McCallModelContinuous(NamedTuple):
+    c: float            # unemployment compensation
+    β: float            # discount factor
+    σ: float            # scale parameter in lognormal distribution
+    μ: float            # location parameter in lognormal distribution
+    w_draws: jnp.ndarray  # draws of wages for Monte Carlo
+
```
Comment (HumphreyYang) [inline lectures/mccall_model.md:778]:
```suggestion
class McCallModelContinuous(NamedTuple):
    c: float              # unemployment compensation
    β: float              # discount factor
    σ: float              # scale parameter in lognormal distribution
    μ: float              # location parameter in lognormal distribution
    w_draws: jnp.ndarray  # draws of wages for Monte Carlo

```

File: lectures/mccall_model.md [hunk 8]
```diff
@@ -789,48 +770,41 @@ Once your code is working, investigate how the reservation wage changes with $c$
...
+def create_mccall_continuous(c=25, β=0.99, σ=0.5, μ=2.5, mc_size=1000, seed=1234):
```
Comment (HumphreyYang) [inline lectures/mccall_model.md:779]: Similar `seed` change

```suggestion
def create_mccall_continuous(c=25, β=0.99, σ=0.5, μ=2.5, mc_size=1000, seed=0):
```

File: lectures/mccall_model.md [hunk 9]
```diff
@@ -840,15 +814,17 @@ We will do this using a contour plot.
...
+    mcmc = create_mccall_continuous(c=c, β=β)
+    return compute_reservation_wage_continuous(mcmc)
 
-for i, c in enumerate(c_vals):
-    for j, β in enumerate(β_vals):
-        mcmc = McCallModelContinuous(c=c, β=β)
-        R[i, j] = compute_reservation_wage_continuous(mcmc)
+# Create meshgrid and vectorize computation
+c_grid, β_grid = jnp.meshgrid(c_vals, β_vals, indexing='ij')
+compute_R_vectorized = jax.vmap(jax.vmap(compute_R_element, in_axes=(None, 0)), in_axes=(0, None))
```
Comment (HumphreyYang) [inline lectures/mccall_model.md]: Just to keep it under 80 characters:

```suggestion
compute_R_vectorized = jax.vmap(
    jax.vmap(compute_R_element, 
    in_axes=(None, 0)), 
    in_axes=(0, None))
```

**Review by HumphreyYang — CHANGES_REQUESTED** [review]
Many thanks @longye-tian! Nice changes!!

Just some minor suggestions in the review.

Three minor general suggestions:

1. There are some comments like this in the code:
```
    # == Now compute the reservation wage == #
```
It would be nice to convert them to more standard comments like 
```
# Now compute the reservation wage
```
2. We used `jax.random` many times. It might be a good idea to give it an alias in the import:
```python
import jax.random as jr
```
and then use 
```python
jr.method_name()
```

3. There are a few more lines that are longer than the box in the preview. It would be nice if you could follow the pattern in the review comment and cut them to fewer than 80 characters so they fit into the code cell in the preview nicely!

**Comment by longye-tian** [comment]
Hi @HumphreyYang ,

Thank you for your detailed review. I've updated the lecture accordingly.

Best,
Longye

File: lectures/mccall_model.md (hunk 1 above)
Comment (longye-tian) [reply lectures/mccall_model.md:62]: Thanks Humphrey.

File: lectures/mccall_model.md (hunk 3 above)
Comment (longye-tian) [reply lectures/mccall_model.md]: Thanks Humphrey. Ive deleted it

File: lectures/mccall_model.md (hunk 4 above)
Comment (longye-tian) [reply lectures/mccall_model.md]: Thanks Humphrey. Changed

File: lectures/mccall_model.md
Comment (shlff) [inline lectures/mccall_model.md:1]: an issue in line 143: 

> {eq}`objective` when an agent in this situation makes *optimal* decisions now 

issue: there is no equation labeled as `objective` and the reference {eq}`objective` directs to a equation (43.1) in another lecture [mccall_model_with_separation](https://pr-619--sunny-cactus-210e3e.netlify.app/mccall_model_with_separation#equation-objective). 

suggestion: label the equation (\mathbb{E} \sum_{t=0}^{\infty} \beta^t y_t) from lines 96-98 of the current lecture as `objective` and cite it from there.

Comment (longye-tian) [reply lectures/mccall_model.md:1]: Thank you @shlff .

I updated the lecture accordingly

**Comment by longye-tian** [comment]
Hi @mmcky ,

I've updated the lecture and revised according to very good suggestions from @HumphreyYang and @shlff .

What do you think?

Best,
Longye

**Comment by jstac** [comment]
@mmcky FYI, I'm getting a 404 on the netlify link above.  

I'll probably push some changes now, which might generate a fresh link that works.

@longye-tian @HumphreyYang Actually I disagree with `import jax.random as jr`, although I appreciate the suggestion.  The only reason is that it's nonstandard, so requires a search if someone is trying to work out what it means.  

I'll revert the change now and review the lecture.

**Comment by mmcky** [comment]
> @mmcky FYI, I'm getting a 404 on the netlify link above.

thanks for reporting @jstac. I can't seem to replicate this on my end. The recent ones appear to work for me. Would you mind pasting the link that is 404'ing for you?

**Comment by HumphreyYang** [comment]
> @longye-tian @HumphreyYang Actually I disagree with `import jax.random as jr`, although I appreciate the suggestion. The only reason is that it's nonstandard, so requires a search if someone is trying to work out what it means.

Many thanks @jstac, I think there is an entry in the style guide that suggests us to use `import jax.random as jr` so I proposed the change in this PR:
https://manual.quantecon.org/styleguide/jax-conversion.html#random-number-generation

I have opened a PR on this in the style sheet:
https://github.com/QuantEcon/QuantEcon.manual/pull/76

**Comment by jstac** [comment]
> thanks for reporting @jstac. I can't seem to replicate this on my end. The recent ones appear to work for me. Would you mind pasting the link that is 404'ing for you?

All links working again after my last push.  Do they go stale after a while?

> Many thanks @jstac, I think there is an entry in the style guide that suggests us to use import jax.random as jr so I proposed the change in this PR:
https://manual.quantecon.org/styleguide/jax-conversion.html#random-number-generation

Ah, thanks for letting me know @HumphreyYang 

I have opened a PR on this in the style sheet:
https://github.com/QuantEcon/QuantEcon.manual/pull/76

Nice work.

**Comment by jstac** [comment]
Please hold off on any edits to this PR.

**Comment by jstac** [comment]
🤖 **Status note for a future session** — from a maintainer investigation on 2026-07-08 into why open-PR previews 404. Context only, not instructions.

**Netlify preview:** https://pr-619--sunny-cactus-210e3e.netlify.app/ currently returns **404**.

### Why previews are down (repo-wide findings)

**1. This branch is stale — 178 commits behind `main`.** A preview build compiles the *whole site from this branch*. This branch's `lectures/house_auction.md` still has unpinned `!pip install prettytable`, which now breaks on a `wcwidth` incompatibility. `main` fixed this on 2026-06-28 by pinning `prettytable<3.18` (#939). **This alone fails any rebuild of this branch until it's updated to `main`.**

**2. The arviz failure was a red herring — do NOT pin arviz or rewrite plotting.** A 2026-07-07 rebuild also failed in `ar1_bayes`/`ar1_turningpts` with an `arviz_plots` `figsize` ValueError. That was a *transient bug in an intermediate arviz-plots 1.x release, already fixed in arviz 1.2.0*. Verified locally on a clean latest-stack venv: the real `az.plot_trace(trace)` cell (pymc + numpyro InferenceData) runs green. The lectures use only 1.x-compatible arviz APIs (`plot_trace`, `summary`, `from_numpyro`, `compare`).

### Recommended first step for this PR

Update this branch to `main` (merge or rebase — pulls in #939 plus ~178 other commits), then let CI rebuild. On today's latest libraries the site builds clean, so the preview should return. `house_auction` is the *known* blocker; updating also picks up other since-merged fixes — rebuild and address any remaining per-lecture failures. Verify with:

```
curl -sI https://pr-619--sunny-cactus-210e3e.netlify.app/mccall_model.html
```

---
*This PR touches: `mccall_model.md`. Last CI build: `success@2025-10-31`. Branch: 178 commits behind `main` as of 2026-07-08.*
